### ISSUE TITLE
Resize API is improperly supported

### ISSUE THREAD

**Issue by baconpaul** [free-audio/clap-juce-extensions#56]
What was I thinking when I wrote this? 

https://github.com/free-audio/clap-juce-extensions/blob/326c380d1f737ac255a377e5ad7b300b61864b3f/src/wrapper/clap-juce-wrapper.cpp#L874-L884

it happens to work surge mac but is totally off.

**Comment by baconpaul** [comment]
OK so here's where we are with bws 4.3 b9 (and tagging in @abique here)

I think it might be impossible to write guiAdjustSize stably

If I take clap-juce-wrapper and set the mode to terminate / maximal using the undefs at the top

and then replace the adjust size with this

```cpp
            // width = std::round(aspectRatio * height);
            bool adjustWidth = (1.0 * width / height > aspectRatio);

            if (adjustWidth)
            {
                width = (uint32_t)std::round(1.0 * height * aspectRatio);
            }
            else
            {
                height = (uint32_t)std::round(1.0 * width / aspectRatio);
            }
```

in surge I get this

```
clap_plugin_gui.adjust_size() isn't stable:
  (817, 569) -> (817, 559)
  (817, 559) -> (816, 559)
  !! Check you're rounding math!
libc++abi: terminating
```

In surge the aspect ratio is  is 1.46042

So lets do the math.

817 / 569 -> 1.435852 so that is < aspect ratio. so adjust height

height -> 817 / 1.46042 = 559.4281 rounds to 559. Great

Then lets put in 817, 559

817 / 559 = 1.462153 which is > aspect ratio

so width goes to 559 * 1.46042 = 816.374 round to 816

in other words, that simple algorithm isn't stable under rounding.

This makes bitwig throw an exception and blammo.

**Comment by baconpaul** [comment]
If I add to the end of that calculation

```cpp
 // never do more less than a 2px adjustment
            if (abs((int)width - (int)*w) < 2 && abs((int)height - (int)*h) < 2)
            {
                width = *w;
                height = *h;
            }
```

then it no longer crashes. Namely clamp my changes always.

(It then only ever shrinks with side and bottom drags but that's a separate question)

**Comment by baconpaul** [comment]
OK I got firther

If you run clap-juce-extensions from `baconpaul/almost-sized` then

1. it is stable
2. left and bottom drag work
3. but corner drag flickers nastily

at least on a mac retina display with surge

i can pick this up in a couple of weeks but if anyone else wants to take a swing I would welcome it.

**Comment by baconpaul** [comment]
Oh and here's the code in that branch

```
          auto ch = editor->getBounds().getHeight();
            auto cw = editor->getBounds().getWidth();

            auto dw = (int)width - (int)cw;
            auto dh = (int)height - (int)ch;

            bool adjustWidth = true;

            if (abs(dw) > 2 && abs(dh) <= 2)
            {
                adjustWidth = false;
            }
            else if (abs(dh) > 2 && abs(dw) <= 2)
            {
                adjustWidth = true;
            }

            if (adjustWidth)
            {
                width = (uint32_t)std::round(1.0 * height * aspectRatio);
            }
            else
            {
                height = (uint32_t)std::round(1.0 * width / aspectRatio);
            }

            // never do more less than a 2px adjustment
            if (abs((int)width - (int)*w) < 2 && abs((int)height - (int)*h) < 2)
            {
                width = *w;
                height = *h;
            }

            if (abs(dw) <= 1 && abs(dh) <= 1)
            {
                width = *w;
                height = *h;
            }
```
